### PR TITLE
PERF: Meta data dictionary move

### PR DESCRIPTION
**Pull request by hjmjohnson**
http://review.source.kitware.com/#/c/23238/5

Bradley Lowekamp
Uploaded patch set 1.Mar 5, 2018

Bradley Lowekamp
Patch Set 1: Straight from SimpleITK, what do you think of this idea?Mar 5, 2018

Kitware Build Robot
Patch Set 1: Verified-1 Build Failed: CDash filtered results: https://open.cdash.org/index.php?&project=Insight&filtercount=3&field1=buildname/string&compare1=63&value1=23238-1&field2=buildstarttime/date&compare2=83&value2=2015-03-01&field3=buildstarttime/date&compare3=84&value3=2029-01-01Mar 5, 2018

Dzenan Zukic
Patch Set 1: (2 comments) Looks good, but as you said a lot of testing is needed.Mar 6, 2018

Bradley Lowekamp
Patch Set 1: (1 comment) Thanks for the feed back.Mar 6, 2018

Bradley Lowekamp
Uploaded patch set 2.Mar 6, 2018

Dzenan Zukic
Patch Set 2: Code-Review+1Mar 6, 2018

Kitware Build Robot
Patch Set 2: Verified-1 Build Failed: CDash filtered results: https://open.cdash.org/index.php?&project=Insight&filtercount=3&field1=buildname/string&compare1=63&value1=23238-2&field2=buildstarttime/date&compare2=83&value2=2015-03-01&field3=buildstarttime/date&compare3=84&value3=2029-01-01Mar 6, 2018

Bradley Lowekamp
Uploaded patch set 3.May 7, 2018

Kitware Build Robot
Patch Set 3: Verified-1 Build Failed: CDash filtered results: https://open.cdash.org/index.php?&project=Insight&filtercount=3&field1=buildname/string&compare1=63&value1=23238-3&field2=buildstarttime/date&compare2=83&value2=2015-03-01&field3=buildstarttime/date&compare3=84&value3=2029-01-01May 7, 2018

Bradley Lowekamp
Uploaded patch set 4: Patch Set 3 was rebased.May 25 1:30 PM

Kitware Build Robot
Patch Set 4: Verified+1 Build Successful: CDash filtered results: https://open.cdash.org/index.php?&project=Insight&filtercount=3&field1=buildname/string&compare1=63&value1=23238-4&field2=buildstarttime/date&compare2=83&value2=2015-03-01&field3=buildstarttime/date&compare3=84&value3=2029-01-01May 25 4:08 PM

Bradley Lowekamp
Uploaded patch set 5: Patch Set 4 was rebased.May 29 8:30 AM

Kitware Build Robot
Patch Set 5: Build Successful: CDash filtered results: https://open.cdash.org/index.php?&project=Insight&filtercount=3&field1=buildname/string&compare1=63&value1=23238-5&field2=buildstarttime/date&compare2=83&value2=2015-03-01&field3=buildstarttime/date&compare3=84&value3=2029-01-01May 29 9:46 AM

Bradley Lowekamp
Patch Set 5: Add more comments to ensure that future developers keep the copy on write working correctlyJul 27 10:36 AM

Hans J. Johnson
Patch Set 5: Code-Review+1 This looks like like a nice improvement.Aug 23 10:49 AM

### CODE REVIEW

**Comment by thewtex** [comment]
@hjmjohnson The WIP check looks at the title of the PR.

**Comment by hjmjohnson** [comment]
Accidently deleted/closed branch when the script for moving from Gerrit went haywire.

File: Modules/Core/Common/src/itkMetaDataDictionary.cxx [hunk 1]
```diff
@@ -21,24 +21,28 @@ namespace itk
...
 ::MetaDataDictionary(const MetaDataDictionary & old)
+  :  m_Dictionary( new MetaDataDictionaryMapType(*old.m_Dictionary))
 {
-  m_Dictionary = new MetaDataDictionaryMapType;
-  *m_Dictionary = *( old.m_Dictionary );
 }
 
+MetaDataDictionary
+::MetaDataDictionary( MetaDataDictionary && rr)
+  :  m_Dictionary( new MetaDataDictionaryMapType(std::move(*rr.m_Dictionary)))
```
Comment (N-Dekker) [inline Modules/Core/Common/src/itkMetaDataDictionary.cxx]: I guess you know, dynamic memory allocation (as in `new MetaDataDictionaryMapType`) in a move-operation should _preferably_ be avoided. Wouldn't a defaulted move-constructor (`= default`) just do the right thing?

File: Modules/Core/Common/src/itkMetaDataDictionary.cxx [hunk 2]
```diff
@@ -21,30 +21,46 @@ namespace itk
...
+  :  m_Dictionary( old.m_Dictionary )
 {
-  m_Dictionary = new MetaDataDictionaryMapType;
-  *m_Dictionary = *( old.m_Dictionary );
 }
 
+MetaDataDictionary
+::MetaDataDictionary( MetaDataDictionary && rr)
+  :  m_Dictionary( std::move(rr.m_Dictionary) )
+{
```
Comment (N-Dekker) [inline Modules/Core/Common/src/itkMetaDataDictionary.cxx]: Still looks like explicitly defaulted move-constructor (`= default`) would just do the right thing, OK?

File: Modules/Core/Common/src/itkMetaDataDictionary.cxx (hunk 1 above)
Comment (blowekamp) [reply Modules/Core/Common/src/itkMetaDataDictionary.cxx]: Good catch!

**Comment by blowekamp** [comment]
I think the move methods should be kept explicit. This class is implementing a non trivial resource management policy so I think being explicit is a very good thing.

**Comment by blowekamp** [comment]
Thanks again for your comments and review. I will seek to combine modern C++ style with ITK's tradition of being explicit. I am reviving the PR in #455

**Comment by hjmjohnson** [comment]
Moved to : Meta data dictionary move #455.

**Pull request closed without merging**